# Cancer Diagnosing AI
***

### Imports

In [20]:
import pandas as pd
from sklearn.model_selection import train_test_split
import tensorflow as tf

### Read in data

In [21]:
data = pd.read_csv('tumor_data.csv')

### Define x Features and y Output

In [22]:
# The input x is all of the tumor features
x = data.drop(columns=["diagnosis(1=m, 0=b)"])

In [32]:
x.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   fractal_dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [23]:
# The output y is a classification of whether a cancerous tumor is malignant or beign
y = data["diagnosis(1=m, 0=b)"]

In [33]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: diagnosis(1=m, 0=b), dtype: int64

### Split our Data 80/20 for Training and Testing respectively

In [24]:
# 80% of our data will be for training, 20% for testing
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2)

### Build a Neural Net

In [25]:
model = tf.keras.models.Sequential()

#### Add layers

In [36]:
# Define our input layer, which will output 256 neurons
# Dense is
# Sigmoid func will help to squish neural net values between 0 & 1, those output values will be used for diagnosis
model.add(tf.keras.layers.Dense(256, input_shape = x_train.shape, activation = "sigmoid")) 

# Second layer
model.add(tf.keras.layers.Dense(256, activation = "sigmoid"))

# Output layer
model.add(tf.keras.layers.Dense(1, activation = "sigmoid")) 

### Compile the Model

In [37]:
# Adam fine tunes the weights of the algorithm to fit the data
model.compile(optimizer = "adam", loss = "binary_crossentropy", metrics = ["accuracy"])

In [38]:
model.fit(x_train, y_train, epochs = 1000)

# Using a thousand epochs is overkill however it works with this relatively small dataset
# Also when it comes to cancer better to be safe than sorry!

Epoch 1/1000
15/15 [==============================] - 0s 2ms/step - loss: 0.6917 - accuracy: 0.5560
Epoch 2/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.6772 - accuracy: 0.5560
Epoch 3/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.6449 - accuracy: 0.6505
Epoch 4/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.5428 - accuracy: 0.7824
Epoch 5/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.3136 - accuracy: 0.8967
Epoch 6/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.2324 - accuracy: 0.9099
Epoch 7/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.2524 - accuracy: 0.9121
Epoch 8/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.2560 - accuracy: 0.9011
Epoch 9/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.2455 - accuracy: 0.9121
Epoch 10/1000
15/15 [==============================] - 0s 1ms/step - loss: 0.2376 - accuracy: 0.9055

In [39]:
model.evaluate(x_test,y_test)

4/4 [==============================] - 0s 898us/step - loss: 0.1028 - accuracy: 0.9737


[0.10284560173749924, 0.9736841917037964]

### Success! This neural net can classify malignant/benign cancerous tumors with 97% accuracy :)